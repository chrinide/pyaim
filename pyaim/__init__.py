#!/usr/bin/env python

import os

__version__ = '0.1'

DEBUG = True

__path__.append(os.path.join(os.path.dirname(__file__), 'tools'))

del(os)
