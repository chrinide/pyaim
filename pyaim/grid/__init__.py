from pyaim.grid.angular import lebgrid
