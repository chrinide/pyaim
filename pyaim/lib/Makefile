#!/usr/bin/env make

CC = gcc
LD = gcc
CFLAGS = -O3 -mtune=native -fopenmp -fpic -lm 
CDEBUG = -g -fpic -fopenmp
LFLAGS = -shared -L/home/jluis/src/pyscf/dev/pyscf/lib -lcgto

all: libaim.so

OBJECTS = surf.o 

libaim.so: $(OBJECTS)
	$(LD) $(LFLAGS) -o libaim.so $(OBJECTS)

clean:
	/bin/rm *.so *.o

surf.o: surf.h

.SUFFIXES:
.SUFFIXES: .c .o .f90

.c.o:
	$(CC) -c $(CFLAGS) -o $@ $<

