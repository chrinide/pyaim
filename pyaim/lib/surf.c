/*
 * Atoms in moleucles module
 * Author: Jose Luis Casals Sainz <jluiscasalssainz@gmail.com>
 */

#include <stdio.h>
#include <stdbool.h>
#include <assert.h>
#include <stdlib.h>
#include <math.h>
#include <omp.h>
                   
#include "surf.h"

void surf_driver(int inuc, int npang, double *ct, double *st, 
                 double *cp, double *sp,  
                 int ntrial, double *rpru, double epsiscp,
                 double epsroot, double rmaxsurf, int backend,
                 double epsilon, double step, int mstep,
                 int cart, double *coord, int *atm, int natm, 
                 int *bas, int nbas, double *env, int nprim,
                 int *ao_loc,
                 double *mo_coeff, double *mo_occ, 
                 int *nlimsurf, double *rsurf){

  int i, j;

  // Setup surface info
  inuc_ = inuc;
  natm_ = natm;
  coord_ = coord;
  epsiscp_ = epsiscp;
  ntrial_ = ntrial;      
  npang_ = npang;
  epsroot_ = epsroot;
  rmaxsurf_ = rmaxsurf;
  backend_ = backend;
  epsilon_ = epsilon;
  step_ = step;
  mstep_ = mstep;
	rpru_ = (double *) malloc(sizeof(double)*ntrial_);
  assert(rpru_ != NULL);
  for (i=0; i<ntrial_; i++){
    rpru_[i] = rpru[i];
  }
	rsurf_ = (double *) malloc(sizeof(double)*npang_*ntrial_);
  assert(rsurf_ != NULL);
	nlimsurf_ = (int *) malloc(sizeof(int)*npang_);
  assert(nlimsurf_ != NULL);
	ct_ = (double *) malloc(sizeof(double)*npang_);
  assert(ct_ != NULL);
  st_ = (double *) malloc(sizeof(double)*npang_);
  assert(st_ != NULL);
	cp_ = (double *) malloc(sizeof(double)*npang_);
  assert(cp_ != NULL);
	sp_ = (double *) malloc(sizeof(double)*npang_);
  assert(sp_ != NULL);
  for (i=0; i<npang_; i++){
    ct_[i] = ct[i];
    st_[i] = st[i];
    cp_[i] = cp[i];
    sp_[i] = sp[i];
  }
	xnuc_ = (double *) malloc(sizeof(double)*3);
  assert(xnuc_ != NULL);
  xnuc_[0] = coord_[inuc_*3+0];
  xnuc_[1] = coord_[inuc_*3+1]; 
  xnuc_[2] = coord_[inuc_*3+2];

  // Basis info
  nprim_ = nprim;
  nbas_= nbas;
  cart_ = cart;
  atm_ = atm;
  bas_ = bas;
  env_ = env;
  ao_loc_ = ao_loc;
	non0tab_ = (int8_t *) malloc(sizeof(int8_t)*nbas_);
  assert(xnuc_ != NULL);
  for (i=0; i<nbas_; i++){
    non0tab_[i] = 1.0;
  }
	shls_ = (int *) malloc(sizeof(int)*2);
  assert(shls_ != NULL);
  shls_[0] = 0;
  shls_[1] = nbas_;
  nmo_ = 0;
  for (i=0; i<nprim_; i++){
    if (mo_occ[i] != 0) nmo_ += 1;
  }
	mo_coeff_ = (double *) malloc(sizeof(double)*nmo_*nprim_);
  assert(mo_coeff_ != NULL);
	mo_occ_ = (double *) malloc(sizeof(double)*nmo_);
  assert(mo_occ_ != NULL);
  int k = 0;
	for (i=0; i<nprim_; i++){ // Orbital
    if (mo_occ[i] != 0){
      mo_occ_[k] = mo_occ[i];
		  for (j=0; j<nprim_; j++){
        mo_coeff_[k*nprim_+j] = mo_coeff[j*nprim_+i];
      }
      k += 1;
    }
	}

  surface();
	for (i=0; i<npang_; i++){
    nlimsurf[i] = nlimsurf_[i];
	  for (j=0; j<ntrial_; j++){
      rsurf[i*ntrial+j] = rsurf_[i*ntrial_+j];
    }
  }

  free(mo_coeff_);
  free(mo_occ_);
  free(coord_);
  free(atm_);
  free(bas_);
  free(env_);
  free(rpru_);
  free(cp_);
  free(sp_);
  free(ct_);
  free(st_);
  free(xnuc_);
  free(non0tab_);
  free(shls_);
  free(ao_loc_);

}

void rhograd(double *point, double *rho, double *grad, double *gradmod){

	double ao_[nprim_*4];
  double c0_[nmo_],c1_[nmo_],c2_[nmo_],c3_[nmo_];

  if (cart_ == 1) {
    aim_GTOval_cart_deriv1(1, shls_, ao_loc_, ao_, point, 
                       non0tab_, atm_, natm_, bas_, nbas_, env_);
  }
  else {
    aim_GTOval_sph_deriv1(1, shls_, ao_loc_, ao_, point, 
                      non0tab_, atm_, natm_, bas_, nbas_, env_);
  }

  int i, j;

  for (i=0; i<nmo_; i++){
    c0_[i] = 0.0;
    c1_[i] = 0.0;
    c2_[i] = 0.0;
    c3_[i] = 0.0;
    for (j=0; j<nprim_; j++){
      c0_[i] += ao_[j+nprim_*0]*mo_coeff_[i*nprim_+j];
      c1_[i] += ao_[j+nprim_*1]*mo_coeff_[i*nprim_+j];
      c2_[i] += ao_[j+nprim_*2]*mo_coeff_[i*nprim_+j];
      c3_[i] += ao_[j+nprim_*3]*mo_coeff_[i*nprim_+j];
    }
  }

  *rho = 0.0;
  grad[0] = 0.0;
  grad[1] = 0.0;
  grad[2] = 0.0;
  *gradmod = 0.0;

  for (i=0; i<nmo_; i++){
    *rho += c0_[i]*c0_[i]*mo_occ_[i];
    grad[0] += c1_[i]*c0_[i]*mo_occ_[i]*2.0;
    grad[1] += c2_[i]*c0_[i]*mo_occ_[i]*2.0;
    grad[2] += c3_[i]*c0_[i]*mo_occ_[i]*2.0;
  }

  *gradmod = grad[0]*grad[0];
  *gradmod += grad[1]*grad[1];
  *gradmod += grad[2]*grad[2];
  *gradmod = sqrt(*gradmod);
  grad[0] = grad[0]/(*gradmod + HMINIMAL);
  grad[1] = grad[1]/(*gradmod + HMINIMAL);
  grad[2] = grad[2]/(*gradmod + HMINIMAL);

}

bool checkcp(const double *x, const double rho, const double gradmod, int *nuc){

  int i;
  bool iscp = false;
  *nuc = -2;

  for (i=0; i<natm_; i++){
    double r = 0.0;
    r =  (x[0]-coord_[i*3+0])*(x[0]-coord_[i*3+0]);
    r += (x[1]-coord_[i*3+1])*(x[1]-coord_[i*3+1]);
    r += (x[2]-coord_[i*3+2])*(x[2]-coord_[i*3+2]);
    r = sqrt(r);
    if (r <= epsiscp_){
      iscp = true;
      *nuc = i;
      return iscp;
    }
  }

  if (gradmod <= GRADEPS){
    iscp = true;
    if (rho <= RHOEPS){
      *nuc = -1;
    }
  } 

  return iscp;
  
}

//Runge-Kutta-Cash-Karp
void stepper_rkck(const double *xpoint, const double *grdt, const double h0, double *xout, double *xerr){

  static const double b21 = 1.0/5.0;
  static const double b31 = 3.0/40.0;
  static const double b32 = 9.0/40.0;
  static const double b41 = 3.0/10.0;
  static const double b42 = -9.0/10.0;
  static const double b43 = 6.0/5.0;
  static const double b51 = -11.0/54.0;
  static const double b52 = 5.0/2.0;
  static const double b53 = -70.0/27.0;
  static const double b54 = 35.0/27.0;
  static const double b61 = 1631.0/55296.0;
  static const double b62 = 175.0/512.0;
  static const double b63 = 575.0/13824.0;
  static const double b64 = 44275.0/110592.0;
  static const double b65 = 253.0/4096.0;
  static const double c1 = 37.0/378.0;
  static const double c3 = 250.0/621.0;
  static const double c4 = 125.0/594.0;
  static const double c6 = 512.0/1771.0;
  static const double dc1 = c1-(2825.0/27648.0);
  static const double dc3 = c3-(18575.0/48384.0);
  static const double dc4 = c4-(13525.0/55296.0);
  static const double dc5 = -277.0/14336.0;
  static const double dc6 = c6-(1.0/4.0);
  
  double rho, grad[3], gradmod;

  xout[0] = xpoint[0] + h0*b21*grdt[0];
  xout[1] = xpoint[1] + h0*b21*grdt[1];
  xout[2] = xpoint[2] + h0*b21*grdt[2];

  double ak2[3] = {0.0};
  rhograd(xout, &rho, grad, &gradmod);
  ak2[0] = grad[0];
  ak2[1] = grad[1];
  ak2[2] = grad[2];
  xout[0] = xpoint[0] + h0*(b31*grdt[0]+b32*ak2[0]);
  xout[1] = xpoint[1] + h0*(b31*grdt[1]+b32*ak2[1]);
  xout[2] = xpoint[2] + h0*(b31*grdt[2]+b32*ak2[2]);
  
  double ak3[3] = {0.0};
  rhograd(xout, &rho, grad, &gradmod);
  ak3[0] = grad[0]; 
  ak3[1] = grad[1]; 
  ak3[2] = grad[2]; 
  xout[0] = xpoint[0] + h0*(b41*grdt[0]+b42*ak2[0]+b43*ak3[0]);
  xout[1] = xpoint[1] + h0*(b41*grdt[1]+b42*ak2[1]+b43*ak3[1]);
  xout[2] = xpoint[2] + h0*(b41*grdt[2]+b42*ak2[2]+b43*ak3[2]);
  
  double ak4[3] = {0.0};
  rhograd(xout, &rho, grad, &gradmod);
  ak4[0] = grad[0];
  ak4[1] = grad[1];
  ak4[2] = grad[2];
  xout[0] = xpoint[0] + h0*(b51*grdt[0]+b52*ak2[0]+b53*ak3[0]+b54*ak4[0]);
  xout[1] = xpoint[1] + h0*(b51*grdt[1]+b52*ak2[1]+b53*ak3[1]+b54*ak4[1]);
  xout[2] = xpoint[2] + h0*(b51*grdt[2]+b52*ak2[2]+b53*ak3[2]+b54*ak4[2]);
  
  double ak5[3] = {0.0};
  rhograd(xout, &rho, grad, &gradmod);
  ak5[0] = grad[0];
  ak5[1] = grad[1];
  ak5[2] = grad[2];
  xout[0] = xpoint[0] + h0*(b61*grdt[0]+b62*ak2[0]+b63*ak3[0]+b64*ak4[0]+b65*ak5[0]);
  xout[1] = xpoint[1] + h0*(b61*grdt[1]+b62*ak2[1]+b63*ak3[1]+b64*ak4[1]+b65*ak5[1]);
  xout[2] = xpoint[2] + h0*(b61*grdt[2]+b62*ak2[2]+b63*ak3[2]+b64*ak4[2]+b65*ak5[2]);
  
  double ak6[3] = {0.0};
  rhograd(xout, &rho, grad, &gradmod);
  ak6[0] = grad[0];
  ak6[1] = grad[1];
  ak6[2] = grad[2];
  xout[0] = xpoint[0] + h0*(c1*grdt[0]+c3*ak3[0]+c4*ak4[0]+c6*ak6[0]);
  xout[1] = xpoint[1] + h0*(c1*grdt[1]+c3*ak3[1]+c4*ak4[1]+c6*ak6[1]);
  xout[2] = xpoint[2] + h0*(c1*grdt[2]+c3*ak3[2]+c4*ak4[2]+c6*ak6[2]);

  xerr[0] = h0*(dc1*grdt[0]+dc3*ak3[0]+dc4*ak4[0]+dc5*ak5[0]+dc6*ak6[0]);
  xerr[1] = h0*(dc1*grdt[1]+dc3*ak3[1]+dc4*ak4[1]+dc5*ak5[1]+dc6*ak6[1]);
  xerr[2] = h0*(dc1*grdt[2]+dc3*ak3[2]+dc4*ak4[2]+dc5*ak5[2]+dc6*ak6[2]);

}

bool adaptive_stepper(double *x, const double *grad, double *h){

  int ier = 1;
  bool adaptive = true;
  double xout[3], xerr[3];

  while (ier != 0){
    double nerr = 0.0;
    stepper_rkck(x, grad, *h, xout, xerr);
    nerr += xerr[0]*xerr[0];
    nerr += xerr[1]*xerr[1];
    nerr += xerr[2]*xerr[2];
    nerr = sqrt(nerr)/3.0;
    if (nerr <= epsilon_){
      ier = 0;
      x[0] = xout[0];
      x[1] = xout[1];
      x[2] = xout[2];
      if (nerr < epsilon_/10.0){ 
        *h = fmin(MAXSTEP, ENLARGE*(*h));
      }
    } 
    else {
      double scale = SAFETY*(epsilon_/nerr);
      *h = scale*(*h);
      if (fabs(*h) < MINSTEP){
        adaptive = false;
        return adaptive; 
      }
    }
  }

  return adaptive;

}

//ier = 0 (correct), 1 (short step), 2 (too many iterations), 
//      3 (infty), 4 (nna)
int odeint(double *xpoint, double *rho, double *gradmod){

  int ier = 0, i;
  double h0 = step_;
  double grad[3] = {0.0};

	rhograd(xpoint, rho, grad, gradmod);

  if (*rho <= GRADEPS && *gradmod <= RHOEPS){
    ier = 3;
    return ier;
  }

  for (i=0; i<mstep_; i++){
    double xlast[3];
    xlast[0] = xpoint[0];
    xlast[1] = xpoint[1];
    xlast[2] = xpoint[2];
    bool ok = adaptive_stepper(xpoint, grad, &h0);
    if (ok == false){
      xpoint[0] = xlast[0];
      xpoint[1] = xlast[1];
      xpoint[2] = xlast[2];
      ier = 1;
      //exit(0);
      return ier;
    }
	  rhograd(xpoint, rho, grad, gradmod);
    int nuc;
    bool iscp = checkcp(xpoint, *rho, *gradmod, &nuc);
    if (iscp == true){
      ier = 0;
      return ier;
    }
  }

  double dist = 0.0;
  dist = (xpoint[0]-xnuc_[0])*(xpoint[0]-xnuc_[0]);
  dist += (xpoint[1]-xnuc_[1])*(xpoint[1]-xnuc_[1]);
  dist += (xpoint[2]-xnuc_[2])*(xpoint[2]-xnuc_[2]);
  dist = sqrt(dist);
  if (dist < rmaxsurf_){
    ier = 4;
  } 
  else if (dist >= rmaxsurf_){
    ier = 3;
  }
  else{
    ier = 2;
  }
    
  return ier;

}

/*
int *mat = (int *)malloc(rows * cols * sizeof(int));
Then, you simulate the matrix using
int offset = i * cols + j;
// now mat[offset] corresponds to m(i, j)
for row-major ordering and
int offset = i + rows * j;
// not mat[offset] corresponds to m(i, j)
*/

void surface(){

  int i,j;
  double xsurf[ntrial_][3]; 
  int isurf[ntrial_][2];
  int nintersec = 0;
  double xpoint[3];

  if (natm_ == 1){  
	  for (i=0; i<npang_; i++){
      nlimsurf_[i] = 1;
      int offset = i*ntrial_;
      rsurf_[offset] = rmaxsurf_;
      return;
    }
  }

#pragma omp parallel default(none)  \
    private(i,nintersec,j,xpoint,xsurf,isurf) \
    shared(npang_,ct_,st_,cp_,sp_,xnuc_,inuc_,rpru_,\
    ntrial_,epsroot_,rsurf_,nlimsurf_,rmaxsurf_)
{
#pragma omp for nowait schedule(dynamic)
	for (i=0; i<npang_; i++){
    nintersec = 0;
    double cost = ct_[i];
    double sintcosp = st_[i]*cp_[i];
    double sintsinp = st_[i]*sp_[i];
    int ia = inuc_, ib;
    double ra = 0.0, rb;
    double rho, gradmod;
	  for (j=0; j<ntrial_; j++){
      double ract = rpru_[j];
      xpoint[0] = xnuc_[0] + ract*sintcosp; 
      xpoint[1] = xnuc_[1] + ract*sintsinp; 
      xpoint[2] = xnuc_[2] + ract*cost;     
      //TODO: Check for error
      int ier = odeint(xpoint, &rho, &gradmod);
      bool good = checkcp(xpoint, rho, gradmod, &ib);
      rb = ract;
      if (ib != ia && (ia == inuc_ || ib == inuc_)){
        if (ia != inuc_ || ib != -1){
          nintersec += 1;
          xsurf[nintersec-1][0] = ra;
          xsurf[nintersec-1][1] = rb;
          isurf[nintersec-1][0] = ia;
          isurf[nintersec-1][1] = ib;
        }                             
      }
      ia = ib;
      ra = rb;
    }
	  for (j=0; j<nintersec; j++){
      ia = isurf[j][0];
      ib = isurf[j][1];
      ra = xsurf[j][0];
      rb = xsurf[j][1];
      double xin[3], xfin[3];
      xin[0] = xnuc_[0] + ra*sintcosp;
      xin[1] = xnuc_[1] + ra*sintsinp;
      xin[2] = xnuc_[2] + ra*cost;
      xfin[0] = xnuc_[0] + rb*sintcosp;
      xfin[1] = xnuc_[1] + rb*sintsinp;
      xfin[2] = xnuc_[2] + rb*cost;
      while (fabs(ra-rb) > epsroot_){
        double xmed[3];
        xmed[0] = 0.5*(xfin[0] + xin[0]);   
        xmed[1] = 0.5*(xfin[1] + xin[1]);   
        xmed[2] = 0.5*(xfin[2] + xin[2]);   
        double rm = 0.5*(ra + rb);
        xpoint[0] = xmed[0];
        xpoint[1] = xmed[1];
        xpoint[2] = xmed[2];
        int im;
        int ier = odeint(xpoint, &rho, &gradmod);
        bool good = checkcp(xpoint, rho, gradmod, &im);
        if (im == ia){
          xin[0] = xmed[0];
          xin[1] = xmed[1];
          xin[2] = xmed[2];
          ra = rm;
        }
        else if (im == ib){
          xfin[0] = xmed[0];
          xfin[1] = xmed[1];
          xfin[2] = xmed[2];
          rb = rm;
        }
        else{
          if (ia == inuc_){
            xfin[0] = xmed[0];
            xfin[1] = xmed[1];
            xfin[2] = xmed[2];
            rb = rm;
          }
          else{
            xin[0] = xmed[0];
            xin[1] = xmed[1];
            xin[2] = xmed[2];
            ra = rm;
          }
        }
      }
      xsurf[j][2] = 0.5*(ra + rb);
    }
    // organize pairs
    nlimsurf_[i] = nintersec; 
	  for (j=0; j<nintersec; j++){
      int offset = i*ntrial_+j;
      rsurf_[offset] = xsurf[j][2];
    }
    if (nintersec%2 == 0){
      nintersec += 1;
      nlimsurf_[i] = nintersec;
      int offset = i*ntrial_+(nintersec-1);
      rsurf_[offset] = rmaxsurf_;
    }
    //printf("#* %d %d %.6f %.6f %.6f %.6f ",i,nintersec,ct_[i],st_[i],cp_[i],sp_[i]);
	  //for (j=0; j<nintersec; j++){
    // printf(" %.6f ",rsurf_[i*ntrial_+j]);
    //}
    //printf("\n");
  }
}

}

