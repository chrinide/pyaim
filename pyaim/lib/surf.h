/*
 * Atoms in moleucles module
 * Author: Jose Luis Casals Sainz <jluiscasalssainz@gmail.com>
 */

#include <float.h>
#define MAX_THREADS 256

// Atm and basis info
double *coord_;
int *atm_;
int natm_;
int *bas_;
int nbas_;
double *env_;
int cart_;
double *mo_coeff_;
double *mo_occ_;
int nprim_;
int nmo_;
int8_t *non0tab_;
int *shls_;
int *ao_loc_;

// Surface info
#define EPS 1e-7
#define GRADEPS 1e-10
#define RHOEPS 1e-10
#define MINSTEP 1e-4
#define MAXSTEP 0.75
#define SAFETY 0.8
#define HMINIMAL DBL_EPSILON
#define ENLARGE 1.2
int inuc_;        
double epsiscp_;
int ntrial_;      
int npang_;
double epsroot_;
double rmaxsurf_;
int backend_;
double epsilon_;
double step_;
int mstep_;
double xnuc_[3];
double *rpru_;
double *ct_;
double *st_;
double *cp_;
double *sp_;
int *nlimsurf_;
double *rsurf_;

// Functions
void rhograd(double *point, double *rho, double *grad, double *gradmod);
bool checkcp(const double *x, const double rho, const double gradmod, int *nuc);
void stepper_rkck(const double *xpoint, const double *grdt, const double h0, double *xout, double *xerr);
bool adaptive_stepper(double *x, const double *grad, double *h);
int odeint(double *xpoint, double *rho, double *gradmod);

// Drivers
void surface();
void surf_driver(int inuc, int npang, double *ct, double *st, 
                 double *cp, double *sp,  
                 int ntrial, double *rpru, double epsiscp,
                 double epsroot, double rmaxsurf, int backend,
                 double epsilon, double step, int mstep,
                 int cart, double *coord, const double *xyzrho,
                 int *atm, int natm, 
                 int *bas, int nbas, double *env, int nprim,
                 int *ao_loc,
                 double *mo_coeff, double *mo_occ, int *nlimsurf, double *rsurf);

// AO evaluators
void aim_GTOval_sph_deriv1(int ngrids, int *shls_slice, int *ao_loc,
                       double *ao, double *coord, char *non0table,
                       int *atm, int natm, int *bas, int nbas, double *env);
void aim_GTOval_cart_deriv1(int ngrids, int *shls_slice, int *ao_loc,
                       double *ao, double *coord, char *non0table,
                       int *atm, int natm, int *bas, int nbas, double *env);
