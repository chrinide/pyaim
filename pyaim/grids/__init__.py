#!/usr/bin/env python

from pyaim import *
