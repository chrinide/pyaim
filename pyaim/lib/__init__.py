from pyaim.lib import utils
from pyaim.lib.utils import *
