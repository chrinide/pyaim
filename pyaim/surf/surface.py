#!/usr/bin/env python

import os
import sys
import time
import numpy
import ctypes
import h5py

from pyscf import lib
from pyscf.lib import logger
libcgto = lib.load_library('libcgto')

from pyaim.grids import lebgrid

# For code compatiblity in python-2 and python-3
if sys.version_info >= (3,):
    unicode = str

EPS = 1e-7
OCCDROP = 1e-12
GRADEPS = 1e-10
RHOEPS = 1e-6
MINSTEP = 1e-5
MAXSTEP = 0.75
SAFETY = 0.8
ENLARGE = 1.2
HMINIMAL = numpy.finfo(numpy.float64).eps

class BaderSurf(lib.StreamObject):

    def __init__(self, datafile):
        self.inuc = 0
        self.epsiscp = 0.180
        self.ntrial = 11
        self.npang = 5810
        self.epsroot = 1e-4
        self.rmaxsurf = 10.0
        self.rprimer = 0.4
        self.backend = 'rkck'
        self.epsilon = 1e-4 
        self.step = 0.1
        self.mstep = 100
        self.surfile = 'surface.h5'
        self.verbose = logger.NOTE
        self.stdout = sys.stdout
        self.max_memory = lib.param.MAX_MEMORY
        self.chkfile = datafile
        self.scratch = lib.param.TMPDIR 
##################################################
# don't modify the following attributes, they are not input options
        self.mo_coeff = None
        self.mo_occ = None
        self.natm = None
        self.coords = None
        self.charges = None
        self.xnuc = None
        self.xyzrho = None
        self.rpru = None
        self.grids = None
        self.rsurf = None
        self.nlimsurf = None
        self.rmin = None
        self.rmax = None
        self.nelectron = None
        self.charge = None
        self.spin = None
        self.atm = None
        self.bas = None
        self.nbas = None
        self.env = None
        self.ao_loc = None
        self.shls_slice = None
        self.nao = None
        self.non0tab = None
        self.cart = None
        self._keys = set(self.__dict__.keys())

    def dump_input(self):

        if self.verbose < logger.INFO:
            return self
        logger.info(self,'')
        logger.info(self,'******** %s flags ********', self.__class__)
        logger.info(self,'Verbose level %d' % self.verbose)
        logger.info(self,'Scratch dir %s' % self.scratch)
        logger.info(self,'Input data file %s' % self.chkfile)
        logger.info(self,'Surface file %s' % self.surfile)
        logger.info(self,'max_memory %d MB (current use %d MB)',
                 self.max_memory, lib.current_memory()[0])
        logger.info(self,'Surface for nuc %d' % self.inuc)
        logger.info(self,'Nuclear coordinate %.5f  %.5f  %.5f', *self.xnuc)
        logger.info(self,'Rmaxsurface %.5f' % self.rmaxsurf)
        logger.info(self,'Npang points %d' % self.npang)
        logger.info(self,'Ntrial %d' % self.ntrial)
        logger.info(self,'Rprimer %.5f' % self.rprimer)
        logger.debug(self, 'Rpru : %s' % self.rpru) 
        logger.info(self,'Epsiscp %.5f' % self.epsiscp)
        logger.info(self,'Epsroot %.5f' % self.epsroot)
        logger.info(self,'ODE solver %s' % self.backend)
        logger.info(self,'ODE tool %.5f' % self.epsilon)
        logger.info(self,'Max steps in ODE solver %d' % self.mstep)
        logger.info(self,'Num atoms %d' % self.natm)
        logger.info(self,'Num electrons %d' % self.nelectron)
        logger.info(self,'Total charge %d' % self.charge)
        logger.info(self,'Spin %d ' % self.spin)
        logger.info(self,'Atom Coordinates (Bohr)')
        for i in range(self.natm):
            logger.info(self,'Nuclei %d with charge %d position : %.5f  %.5f  %.5f', i, 
                        self.charges[i], *self.coords[i])
        return self

    def build(self):

        t0 = time.clock()
        lib.logger.TIMER_LEVEL = 5

        mol = lib.chkfile.load_mol(self.chkfile)
        self.nelectron = mol.nelectron 
        self.charge = mol.charge    
        self.spin = mol.spin      
        self.natm = mol.natm		
        self.atm = numpy.asarray(mol._atm, dtype=numpy.int32, order='C')
        self.bas = numpy.asarray(mol._bas, dtype=numpy.int32, order='C')
        self.nbas = self.bas.shape[0]
        self.env = numpy.asarray(mol._env, dtype=numpy.double, order='C')
        self.ao_loc = mol.ao_loc_nr()
        self.shls_slice = (0, self.nbas)
        sh0, sh1 = self.shls_slice
        self.nao = self.ao_loc[sh1] - self.ao_loc[sh0]
        self.non0tab = numpy.ones((1,self.nbas), dtype=numpy.int8)
        self.coords = numpy.asarray([(numpy.asarray(atom[1])).tolist() for atom in mol._atom])
        self.charges = mol.atom_charges()
        self.mo_coeff = lib.chkfile.load(self.chkfile, 'scf/mo_coeff')
        self.mo_occ = lib.chkfile.load(self.chkfile, 'scf/mo_occ')
        self.cart = mol.cart

        if (self.ntrial%2 == 0): self.ntrial += 1
        geofac = numpy.power(((self.rmaxsurf-0.1)/self.rprimer),(1.0/(self.ntrial-1.0)))
        self.rpru = numpy.zeros((self.ntrial))
        for i in range(self.ntrial): 
            self.rpru[i] = self.rprimer*numpy.power(geofac,(i+1)-1)
        self.xnuc = self.coords[self.inuc]
        self.rsurf = numpy.zeros((self.npang,self.ntrial))
        self.nlimsurf = numpy.zeros((self.npang), dtype=numpy.int32)
        lebgrid.lebgrid(self.npang)

        if self.verbose >= logger.WARN:
            self.check_sanity()
        if self.verbose > lib.logger.NOTE:
            self.dump_input()

    kernel = build

if __name__ == '__main__':
    name = 'test/n2_rhf.chk'
    surf = BaderSurf(name)
    surf.epsilon = 1e-4
    surf.verbose = 5
    surf.inuc = 0
    surf.kernel()
 
