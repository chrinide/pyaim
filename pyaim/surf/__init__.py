from pyaim.surf.surface import BaderSurf
