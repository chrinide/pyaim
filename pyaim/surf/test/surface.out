
******** <class '__main__.BaderSurf'> flags ********
Verbose level 4
Scratch dir /scratch-ssd/jluis
Input data file test/lif.chk
max_memory 60000 MB (current use 62 MB)
Num atoms 2
Num electrons 12
Total charge 0
Spin 0 
Atom Coordinates (Bohr)
Nuclei 0 with charge 9 position : 0.000000  0.000000  -0.738836
Nuclei 1 with charge 3 position : 0.000000  0.000000  2.216507
Surface file test/lif.chk.h5
Surface for nuc 0
Nuclear coordinate 0.000000  0.000000  -0.738836
Rmaxsurface 10.000000
Npang points 5810
Ntrial 11
Rprimer 0.400000
Epsiscp 0.180000
Epsroot 0.000100
ODE solver rkck
ODE tool 0.000100
Max steps in ODE solver 100
Check rho position 0.000000 0.000000 -0.738836
Rmin for surface 1.811376
Rmax for surface 10.000000
Write HDF5 surface file
Surface of atom 0 saved
